Modernize and document common-hid-packet-parser.js

JoergAtGithub opened the pull request:
- Applied eslint setup changes specific for common-hid-packet-parser.js
- Applied automatic pre-commit changes
- Added missing JSDOC documentation (where applicable copied from https://github.com/mixxxdj/mixxx/wiki/Hid-Packet-Parser-Js )
- Use modern ES6 class syntax

JoergAtGithub committed:
Adapted eslint configuration for common-hid-packet-parser.js
Applied eslint changes for common-hid-packet-parser.js

Swiftb0y commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,31 +1,47 @@
-/** Common HID script debugging function. Just to get logging with 'HID' prefix. */
+/**
+ * Common HID script debugging function. Just to get logging with 'HID' prefix.
+ *
+ * @param message
+ */
```
while you are at it, I think it would make sense to `@deprecate` this function

Swiftb0y commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,31 +1,47 @@
 
-/** Return bit offset based on bitmask */
+/**
+ * Return bit offset based on bitmask
```
```suggestion
 * Get the index of the least significant bit that is 1 in `bitmask`
```

Swiftb0y reviewed (commented):
I don't know how much work you want to put into this. A lot of things could be improved, but I don't want to force you to do too much work. So what scope did you intend for this PR? Just make the new ESLint config pass, improve docs or even modernize large parts the library?

JoergAtGithub commented:
> So what scope did you intend for this PR?

No large changes, only cleanup and one small functional extension to initialize data.

JoergAtGithub committed:
Added and improved JSDOC documentation

JoergAtGithub committed:
Use ES6 class syntax
Some types documented

JoergAtGithub committed:
Documentation fixes

JoergAtGithub committed:
Minor documentation corrections

JoergAtGithub committed:
Added missing definitions/default/comment

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/Main' into hid-parser-update

JoergAtGithub committed:
Added missing type for bit to JSDoc
Removed one duplicate line

daschuer commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1830 @@
+     */
+    getOffset(bitmask) {
+        for (let i = 0; i < 32; i++) {
+            if ((1 & bitmask >> i) !== 0) { return i; }
```
This is hard to understand. Is it 
```suggestion
            if ((bitmask >> i) & 1) { 
                return i; 
            }
```

daschuer commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1830 @@
+     * @param {number} bitmask A bitwise mask of up to 32 bit. All bits set to'1' in this mask are considered.
+     * @returns {number} Index of the least significant bit that is 1 in `bitmask`
+     */
+    getOffset(bitmask) {
```
getLeastSiginficantBitSet() or such?

JoergAtGithub replied:
I aggree, but this would break backward compatibility.

daschuer commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1830 @@
+     * Set Output to toggle between two values. Reset with setOutput(name,'off')
+     *
+     * @param {string} group Defines the group name for the field. The group can be any string, but if it matches a valid Mixxx control group name, it is possible to map a field to a control or output without any additional code.
+     * @param {string} name  Is the name of the control for the field. The name can be any string, but if it matches a valid Mixxx control name in the group defined for field, the system attempts to attach it directly to the correct field. Together group and name form the ID of the field (group.name)
```
looong lines

daschuer commented:
I have left some comments. Not sure what will fit in the PR or should be postponed.

Swiftb0y reviewed (dismissed):
I forgot to actually submit these review comments. Primarily because I've only looked at 10% of code yet. But I guess its better to submit an incomplete review than none at all.

JoergAtGithub commented:
Thanks for the feedback! I will address it!
But some remarks go beyond the scope of this PR:
-Except the class syntax, I changed this code only where it was neccessary to pass the eslint check
-All other changes are automatic formatting by Pre-Commit
-And of cause I documented the code in JSDOC

But I do not plan to reimplement the code itself!

Swiftb0y commented:
> But I do not plan to reimplement the code itself!

I fully understand that. It was not my intend to broaden the scope like that. 
Let my explain my reasoning a bit: When you "modernize" code, only doing syntax and documentation changes is essentially just a mass reformatting. If code looks modern but still employs legacy patterns everywhere, the code is still legacy. No real value has been gained.
So, since we actually want to use modern JS for the new features (and not just for its sleeker syntax), I suggested to actually make use of these features. Also see [the boyscout rule](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)

JoergAtGithub committed:
Made hard understandable code section better readable

JoergAtGithub committed:
Reformated many if statements with line breaks, for better code readability

JoergAtGithub committed:
Marked HIDDebug as deprecated and replaced the also deprecated print in it's implementation by console.log

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1800 @@
+        for (let i = 0; i < 32; i++) {
+            if ((1 & bitmask >> i) !== 0) { return i; }
+        }
+        return 0;
```
With the typescript declarations from #4759 it's really easy to see if a function is available in Mixxx 2.4. The Editor would mark it immediately as error, as you can see this is not the case here:
![grafik](https://user-images.githubusercontent.com/64457745/178114145-2cf64619-2c16-4ee1-9a87-9b7a87a59f81.png)
I also tested this implementation with my mapping and it works flawless.

Swiftb0y replied:
Great. I also tested it on JSBench and its 80% faster on average. Even though that is obviously not the same javascript environment, I'm pretty sure it won't be slower than the current solution.

JoergAtGithub replied:
Any performance improvement helps in such library code. Thank you!
My profiling with Trace showed, that the function where the most time is wasted in this library, is getOutputField - an often called function with many nested loops inside.

Swiftb0y replied:
Yeah makes sense... I'm not familiar enough with the code to see any obvious optimization opportunity. I guess it would be nice if you could add a note to the JSDoc to document that that method is expensive...

JoergAtGithub replied:
Will do! I plan to address this in a follow-up PR - I think some kind of looktable is needed here.

JoergAtGithub committed:
Changed implementation of getOffset, to constant time solution using Math.clz32(), proposed by Niko

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1800 @@
-};
+ *          the reportId parameter instead.
+ */
+class HIDPacket {
+    constructor(name, reportId, callback, header) {
+        this.name = name;
+        this.header = header;
+        this.callback = callback;
+
+        this.reportId = 0;
+        if (reportId !== undefined) {
+            this.reportId = reportId;
+        }
```
I tested this: Your implementation behaves the same as the old one. But omiting the reportID can only work, when also callback and header are omited

Swiftb0y replied:
Right... so I guess the API was designed to be called like this? `HIDPacket("abc", undefined, func, [...])`

JoergAtGithub committed:
Simplify implementation of constructor of HIDPacket

JoergAtGithub committed:
Added todo hint, that the implementation of getOutputField has a performance problem

JoergAtGithub committed:
Use const for signed state of signedPackFormats, and unified this across the code

JoergAtGithub committed:
Replaced variable with a constant

JoergAtGithub committed:
Codespell

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1800 @@
+ * Common HID script debugging function. Just to get logging with 'HID' prefix.
+ *
+ * @param {string} message Message to be printed on controller debug console output
+ */
```
Should I replace the use of HIDDebug by console.log accross this library? IMHO the prefix HID is not needed, since the name of thecontroller is printed before each message.

JoergAtGithub replied:
I replaced all calls of HIDDebug by `console.error`, `console.warn` or `console.log` in the parser itself.

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,1800 @@
+    /**
+     * Add a control bitmask to the HIDBitVector
+     *
+     * @param {string} group Defines the group name for the field. The group can be any string, but if it matches a valid Mixxx control group name, it is possible to map a field to a control without any additional code. Defines the group name for the field. The group can be any string, but if it matches a valid Mixxx control group name, it is possible to map a field to a control or output without ...
+     * @param {string} name Is the name of the control for the field. The name can be any string, but if it matches a valid Mixxx control name in the group defined for field, the system attempts to attach it directly to the correct field. Together group and name form the ID of the field (group.name)
+     * @param {number} bitmask A bitwise mask of up to 32 bit. All bits set to'1' in this mask are considered.
```
I tried this, but couldn't find a JSDoc syntax to use the documentation of the typedef for the parameter. From Doxygen I know @copydoc, but there seems nothing similar in JSDoc.

JoergAtGithub committed:
Reformated code to break long lines

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/Main' into hid-parser-update

JoergAtGithub committed:
Replaced usage of deprecated HIDDebug by console.log, console.warn or console.error depending on the context
Instead of the unnecessary prefix HID, it prints the function name.
Errors and Warnings are marked in the console output anyway, therefore I removed the texts "ERROR" and "WARNING" where they were used.

JoergAtGithub committed:
Made HIDDebug a function of the global-this object to prevent eslint error

JoergAtGithub committed:
Use JSDoc @todo where applicable

JoergAtGithub committed:
Fixed several errors

JoergAtGithub committed:
eslint: Make class instances const

JoergAtGithub committed:
Fixed bug of duplicated variables for scratch ramp

JoergAtGithub committed:
Defined enableScratchCallback

JoergAtGithub committed:
Removed argument, which is not existing in constructor

JoergAtGithub committed:
Added @ts-ignore, to inhibit warnings for duplicate identifier

JoergAtGithub committed:
Corrected type that parseBitVector returns

JoergAtGithub committed:
Replaced wrong eslintrc keyword "readable" by "readonly".
https://eslint.org/docs/latest/user-guide/configuring/language-options#using-configuration-files-1

JoergAtGithub committed:
Adjusted eslint rules for common-hid-packet-parser.js:
-External function print is no longer used
-Disabled camelcase warning, because we can't change the keywords due to backward compatibility

JoergAtGithub committed:
Corrected boolean <-> number conversion

JoergAtGithub committed:
Added guards for safe handling field.value as HIDBitVector

JoergAtGithub committed:
Made variables const

JoergAtGithub committed:
Reduced scope of change variable and initialize it (in one branch it remained undefined, but was used for comparison later)

JoergAtGithub committed:
addControl has no return type -> Do not return undefined

JoergAtGithub committed:
Added missing type and docu for pack

JoergAtGithub committed:
Do not try to initialize field structure again, after usage

JoergAtGithub committed:
Corrected wrong type syntax

JoergAtGithub committed:
Specified missing type

JoergAtGithub committed:
Separated types bitObject / packetField

JoergAtGithub committed:
Use LED-colors as On / Off values in linkOutput

JoergAtGithub committed:
Determine size of output array before initializing it, and initialize it with the final size
Use Uint8Array as type for input and output reports
Use sendOutputReport with Uint8Array buffer as argument

JoergAtGithub committed:
Reduced scope of variables and made them const where possible

JoergAtGithub committed:
Corrected return type of lookupBit

JoergAtGithub committed:
Fixed eslint jsdoc/require-property-description warnings by documenting all class properties

JoergAtGithub committed:
Reverted wrong change of object literal to array, and documented the types accordingly

JoergAtGithub committed:
Defined type alias packetItemId

JoergAtGithub committed:
Corrected definition of callback types

JoergAtGithub committed:
Shortened description of control group and control name arguments
Made the difference between mapped control and control clear

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/Main' into hid-parser-update

JoergAtGithub commented:
> I fully understand that. It was not my intend to broaden the scope like that.
> Let my explain my reasoning a bit: When you "modernize" code, only doing syntax and documentation changes is essentially just a mass reformatting. If code looks modern but still employs legacy patterns everywhere, the code is still legacy. No real value has been gained.

@Swiftb0y I took the extra mile and improved also parts of the code itself, especiall, I fixed eslint and Intellisense warnings and used latest mapping API. But the objective was, that everything remains backward compatible. IMHO, the backward compatibility should be also the main review objective.
The features of this PR are definition of correct types, and full function argument description. This function argument description alone, should be a huge help for any HID mapping developer.
I want not let become this another unreviewable monster PR, therefore I ask you, to not request further additions to this PR.

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/Main' into hid-parser-update

Swiftb0y commented:
Thank you for taking the time to improve the code quality of this PR. It looks very good at first glance. Please understand that reviewing this is already hard since I'm unfamiliar with the pre-existing code. Allow me some more time and feel free to post the occasional reminder ping.

Swiftb0y reviewed (commented):
I don't know when I'll get around to finishing this so I'll just publish me WIP review so someone else can take over if they want.

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
+        const bytes = this.packSizes[field.pack];
+        const signed = this.signedPackFormats.includes(field.pack);
+        if (field.type === "bitvector") {
+            const bitVector = /** @type {HIDBitVector} */ (field.value);
```
This is not the same. The `/** @type {HIDBitVector} */` right of the `=` is a typecast, if it would a line before, it would be a type definition for the constant bitVector.
field.value can be of type number, boolean or HIDBitVector. Because it's verified in the line before, that it's a bitvector, we can safely typecast here.

Swiftb0y replied:
I didn't know typecasts were a thing in JSDoc. Ignore my suggestion.

JoergAtGithub committed:
Replaced "Record." by "Object.", because Record is TypeScript syntax and not valid JSDoc.
--
Unfortunately I had to replace the typedef packetItemId, because the JSDoc Object. syntax for records requires type string as index.
--
For type compatibility with TypeScript (and TypeScript based tools like VS Code), the keyword Object in JSDoc must be upper case. To prevent such tools reporting false positive errors, the setting of eslint-plugin-jsdoc (this case is described here: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/README.md#why-not-capital-case-everything)

JoergAtGithub committed:
Clarify, that "play" is just an example

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
+     *     mapped to a Mixxx control.
+     */
+    pack(data, field) {
+        if (!(field.pack in this.packSizes)) {
+            console.error("HIDPacket.pack - Parsing packed value: invalid pack format " + field.pack);
             return;
         }
```
I think, this is similar, but not equivalent, because the `return;` isn't done by console.assert.

Swiftb0y replied:
I see... I guess `throw`ing in case of a triggered assertion is probably too much of breaking change since mappings rely on the fact that this is handled gracefully?

JoergAtGithub replied:
Yes, my primary objective is to keep everything backward compatible. Without this constraint, not much of the code would remain;-)

JoergAtGithub committed:
Use JSDoc statements for deprecated info of processDelta

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
+        if (group === undefined) {
+            return undefined;
+        }
+        const result = group.match(/\[Channel[0-9]+\]/);
+        if (!result) {
```
I aggree, that one single implementaion would be better, but the return values are different, in case of failure. This returns undefined, while script.deckFromGroup returns 0.

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
-    if (field.type === "bitvector") {
-        // TODO - fix multi byte bit vector outputs
-        if (bytes > 1) {
-            HIDDebug("ERROR: packing multibyte bit vectors not yet supported");
+        this.groups = {};
```
I tried `this.group` instead of `this.groups = {};` but than the script fails at the first access to this.groups with: `TypeError: Type error`

Swiftb0y replied:
I think this is a misunderstanding. I didn't mean to remove the "initialization". I was instead suggesting adding a JSDoc type-annotation for `this.groups`.

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
+     * @returns {HIDPacket} The input packet
+     */
+    getInputPacket(name) {
+        if (!(name in this.InputPackets)) {
+            return undefined;
+        }
+        return this.InputPackets[name];
```
As you write, not possible with today Qt JS Engine

JoergAtGithub committed:
Deprecated unused HIDController.OutputUpdateInterval

JoergAtGithub committed:
Used @default syntax - According the JSDoc spec, the value is optional, I added it nevertheless, because VS Code doesn't print it otherwise

JoergAtGithub committed:
Replaced all strings with embedded expressions by template strings

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
-};
+ * @callback packetCallback
+ * @param {HIDPacket} packet The packet that represents the InputReport
+ * @param {Record.<string, packetField | bitObject>} changed_data The data received from the device
```
Done!

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
-                HIDDebug("ERROR callback provided for " + group + "." + name + " is not a function.");
+                console.error(
+                    "HIDPacket.addControl - Callback provided for " + group + "." + name +
+                    " is not a function.");
```
Done!

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
+        if (deck === undefined) {
+            return undefined;
+        }
+        return "[Channel" + deck + "]";
```
Done!

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
+         * - By default undefined.
+         * - If set, it's a value for timer executed every n ms to update Outputs with updateOutputs()
+         *
+         * @todo This is unused and updateOutputs() doesn't exist - Remove?
```
Done!

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
             }
-        } else {
-            data[index] = (value >> (byte_index * 8)) & 255;
         }
```
Done in #4894 !

JoergAtGithub committed:
Corrected type of packetField.pack
Added type for HIDPacket.packSizes
Added missing information about endianess of pack types

JoergAtGithub committed:
Added missing type/type cast

JoergAtGithub committed:
Added missing @type to HIDPacket.groups

JoergAtGithub commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2310 @@
-    if (packet === undefined) {
-        HIDDebug("ERROR creating modifier: input packet " + this.defaultPacket + " not found");
-        return;
+    /**
+     * Initialize our packet data and callbacks. This does not seem to
+     * work when executed from here, but we keep a stub just in case.
+     */
+    initializePacketData() {}
```
I've no clue. It's the only place where the keyword initializePacketData exist in the Mixxx codebase: https://github.com/mixxxdj/mixxx/search?q=initializePacketData
and this comment is there since the beginnig of the Git history.
=> I will remove it

JoergAtGithub committed:
This was the only place where the keyword initializePacketData exist in the Mixxx codebase, and the comment that it don't work is there since the beginning of the Mixxx Git history.

JoergAtGithub committed:
Tried to explain, what this.timers = {}; is used for.

JoergAtGithub commented:
@Swiftb0y I addressed now all open review comments. I would be glad, if you could confirm, that I did this in your sense.

Swiftb0y commented:
Yes, thank you for your effort. Unfortunately, as I pointed out, I only reviewed the first third. I don't know when or even if I'll get to the rest sometime.

JoergAtGithub commented:
Yes, I fully understand this. Therefore I requested no re-review from you.
But if you aggree, that my changes address your change requests, it would be nice, if you could reset the Changes Requested state.

daschuer reviewed (approved):
I did another review round with my limited JS knowledge. 
And I did not spot obvious issues. 
Given that this a PR just started for documentation purpose, it is already mature. 
Thank you for all the work. 
Can you confirm that this is ready for merge and has been tested with HID Hardware? 

@Swiftb0y and @ywwg Did you had a chance to test this on your hardware? 
If you don't have objections, I will merge this in a few days.
Thank you.

JoergAtGithub commented:
@daschuer Thank you very much for reviewing this!
From my point of view, it's ready to merge! But I've only one HID controller (Traktor Z2 with the mapping from #3038) available for test.
Practical testing with other HID mappings makes definitally sense. To get maximum feedback from users with other controllers, this PR should be merged before the next Beta test cycle.

Swiftb0y commented:
> @Swiftb0y and @ywwg Did you had a chance to test this on your hardware?

I don't have any HID hardware or mapping to test this with. I was only reviewing the code because I have a somewhat determined opinion on how we should handle typing in our JS libraries.

Swiftb0y reviewed (changes requested):
Allow me another couple days to review the rest of the code.

Swiftb0y commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2304 @@
+    get(name) {
+        if (!(name in this.modifiers)) {
+            console.error(`HIDModifierList.get - Unknown modifier: ${name}`);
+            return undefined;
```
the change from `false` to `undefined` could be breaking

Swiftb0y commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -1,1549 +1,2304 @@
+     */
+    add(name) {
+        if (name in this.modifiers) {
+            console.warn(`HIDModifierList.add - Modifier already defined: ${name}`);
```
IIUC you can ask for a stacktrace, thats probably more useful than just printing which function was called. Wdyt?
```suggestion
            console.warn(`Modifier already defined: ${name}`);
            console.trace();
```

JoergAtGithub replied:
I tried console.trace, console.assert and console.exception. All have problems (see #11080).
I suggest to change this later, when these functions work as needed.

Swiftb0y reviewed (changes requested):
little nitpicks. I highly appreciate all the effort put into this code, yet I still don't understand the library even the slightest. In the end I'd call for the deprecation of this library and would favor a smaller library based on `TypedArray`s which integrates into componentsJS. Until then, I highly appreciate all the work you put spend so far. LGTM apart from my two nitpicks.

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/Main' into hid-parser-update

JoergAtGithub committed:
Fixed accidentally changed return value of HIDModifierList.Get in error case

JoergAtGithub commented:
> In the end I'd call for the deprecation of this library and would favor a smaller library based on `TypedArray`s which integrates into componentsJS.

A rewrite from scratch makes sense, but before starting with this task, I would like to improve the C++/JavaScript HID API. Currently all the semantic information, which HID provides about the individual controls of a device, isn't accessible in the JS mapping code.

Swiftb0y reviewed (approved):
Thank you. Sorry for taking so long.

JoergAtGithub commented:
Thank you for the detailed review!
